chain/network/src/routing/graph_v2/mod.rs

```diff
@@ -0,0 +1,668 @@
+}
+
+pub enum NetworkTopologyChange {
+    PeerConnected(PeerId, Edge),
+    PeerDisconnected(PeerId),
+    PeerAdvertisedDistances(network_protocol::DistanceVector),
+}
+
+/// Locally stored properties of a received network_protocol::DistanceVector message
+struct PeerDistances {
+    /// Advertised distances indexed by the local EdgeCache's peer to id mapping.
+    pub distance: Vec<i32>,
+    /// The lowest nonce among all edges used to validate the distances.
+    /// For simplicity, used to expire the entire dstance vector at once.
+    pub min_nonce: u64,
+}
+
+struct Inner {
+    config: GraphConfigV2,
+
+    /// Data structure maintaing information about the entire known network
+    edge_cache: EdgeCache,
+
+    /// Edges of the local node's direct connections
+    local_edges: HashMap<PeerId, Edge>,
+    /// Distances advertised by the local node's direct peers
+    peer_distances: HashMap<PeerId, PeerDistances>,
+
+    /// Distances from the local node to other nodes
+    my_distances: HashMap<PeerId, u32>,
+    /// The latest DistanceVector advertised by the local node
+    my_distance_vector: network_protocol::DistanceVector,
+}
+
+impl Inner {
+    /// Function which verifies signed edges.
+    /// Returns true iff all the edges provided were valid.
+    ///
+    /// This method implements a security measure against an adversary sending invalid edges.
+    /// It verifies edges in parallel until the first invalid edge is found. It adds nonces
+    /// for all the edges verified so far to the cache, but drops all the remaining ones. This way
+    /// the wasted work (verification of invalid edges) is constant, no matter how large the input
+    /// size is.
+    ///
+    /// Edge verification is expensive, and it would be an attack vector if we dropped on the
+    /// floor valid edges verified so far: an attacker could prepare a message containing
+    /// a lot of valid edges, except for the last one, and send it repeatedly to a node.
+    /// The node would then validate all the edges every time, then reject the whole set
+    /// because just the last edge was invalid. Instead, we cache all the edges verified so
+    /// far and return an error only afterwards.
+    #[cfg(not(test))]
+    fn verify_and_cache_edge_nonces(&mut self, edges: &Vec<Edge>) -> bool {
+        metrics::EDGE_UPDATES.inc_by(edges.len() as u64);
+
+        // Collect only those edges which are new to us for verification.
+        let mut unverified_edges = Vec::<Edge>::new();
+        for e in edges {
+            // V2 routing protocol only shares Active edges
+            // TODO(saketh): deprecate tombstones entirely
+            if e.edge_type() != EdgeState::Active {
+                return false;
+            }
+
+            if !self.edge_cache.has_edge_nonce_or_newer(e) {
+                unverified_edges.push(e.clone());
+            }
+        }
+
+        // Verify the new edges in parallel on rayon.
+        // Stop at first invalid edge.
```

Comment:
Is this property provided by rayon itself?

Reply:
It seems we implement it ourselves ([link](https://github.com/near/nearcore/blob/7a3a97a5d1f3489af19e6a5aee27cd094a866168/chain/network/src/concurrency/rayon.rs#L30-L48)). It is a bit confusing that we have our own module named rayon. 🙃